### ISSUE TITLE
Scope/Policy to limit database results to what the user is authorized for

### ISSUE THREAD

**Issue by Roemerb** [spatie/laravel-permission#1186]
Hi,

I have a model in my application for which users have to be specifically authorized to be able to see them and I'm trying to come up with an application-wide method to limit database results for this model to only those that the user is allowed to access. For example, a user would have to have a permission called `view post {id}`, or a permission called `view post` with `post {id}` as the guard name. The user could also have a `view all posts` permission that allows them to view every post. Since I'm using repository classes to access my models, instead of directly, I'd like this to be global for my application. I'd also like pagination to keep working.

I've tried several approaches but I can't really figure it out. I initially tried a policy:

```php
public function view(User $user, Post $post)
{
  // Using direct permissions
  $user->hasPermissionTo('view post ' . $post->id);

  // Using guards
  $user->hasPermissionTo('view post', 'post ' . $post->id);
}
```

This does not work at all for some reason, even though I made sure to register the policy in the `AuthServiceProvider` (bound to the model).

Then I tried a global scope:

```php
class Post extends Model
{
  protected static function boot()
  {
    parent::boot();

    static::addGlobalScope(function (Builder $builder) {
      // This $builder instance is for the `posts` table and has nothing to do with
      // permissions or the currently logged in user.
    });
  }
}
```

As demonstrated above, I don't know how to add a constraint to the `Builder` instance that is provided to limit the results to the results the user is authorized for. So I'm not sure this is the way to go either.

Does anyone have any suggestions on what would be the best way to approach this?
Thanks in advance!

**Comment by dennis-git** [comment]
have you tried in your Policy file

if ($user->hasPermissionTo('view post ')) return $user->id === $post->user_id;

**Comment by MattStrauss** [comment]
@Roemerb 

Sorry to bump an old issue, but I saw the `help wanted` tag and was considering contributing a PR. But before I got started I wanted to make sure that the below was attempted. 

When you wrote the policies, did you return the `bool` value like below? I'm thinking you did and the original code you included in this issue was just a quick example. But if not, then it would make sense that the policy wouldn't work, since it would not be returning anything and therefore would always be falsey. 

```
public function view(User $user, Post $post)
{
  // Using direct permissions
  return $user->hasPermissionTo('view post ' . $post->id);

  // Using guards
  return $user->hasPermissionTo('view post', 'post ' . $post->id);
}
```
